File: dotnet/fs/example/Samples/SampleRunner.fs
// File: SampleRunner.fs
// Project: Samples
// Created Date: 03/02/2023
// Author: Shun Suzuki
// -----
// Last Modified: 12/09/2023
// Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
// -----
// Copyright (c) 2023 Shun Suzuki. All rights reserved.
// 

namespace Samples

open AUTD3Sharp

module SampleRunner =
    let Run (autd : Controller) = 
        let examples = [
                (FocusTest.Test, "Single focus test");
                (BesselBeamTest.Test, "Bessel beam test");
                (PlaneTest.Test, "Plane wave test");
                (WavTest.Test, "Wav modulation test");
                (STMTest.FocusSTMTest, "FocusSTM test");
                (STMTest.GainSTMTest, "GainSTM test");
                (STMTest.SoftwareSTMTest, "SoftwareSTM test");
                (GainHoloTest.Test, "Multiple foci test");
                (CustomTest.Test, "Custom Gain & Modulation test");
                (FlagTest.Test, "Flag test");
                (TransTest.Test, "TransducerTest test")];

        let examples = 
            if autd.Geometry.NumDevices >= 2 then examples @ [(GroupTest.Test, "Group test")] else examples;


        printfn "======== AUTD3 firmware information ========"
        autd.FirmwareInfoList() |> Seq.iter (fun firm -> printfn $"{firm}")
        printfn "============================================"

        let rec run_example () =
            let print_example i =
                let _, name = examples[i] in
                printfn $"[{i}]: {name}"
            [0..examples.Length-1] |> List.iter print_example
            printfn "[Others]: finish"
            
            printf "Choose number: "
            let input = stdin.ReadLine()
            match System.Int32.TryParse input with
                | true,i -> 
                    let f, _ = examples[i] in
                    f(autd)

                    printfn "press any key to finish..."
                    System.Console.ReadKey true |> ignore;

                    printfn "finish."

                    (new Stop()) |> autd.Send |> ignore;

                    run_example()
                | _ -> ()

        run_example()

        autd.Close() |> ignore;
        autd.Dispose() |> ignore;

File: dotnet/cs/example/Samples/SampleRunner.cs
/*
 * File: SampleRunner.cs
 * Project: Samples
 * Created Date: 13/10/2022
 * Author: Shun Suzuki
 * -----
 * Last Modified: 02/12/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2022-2023 Shun Suzuki. All rights reserved.
 * 
 */


using AUTD3Sharp;

namespace Samples;

internal delegate Task TestFn<T>(Controller<T> autd);

public class SampleRunner
{
    public static async Task Run<T>(Controller<T> autd)
    {
        var examples = new List<(TestFn<T>, string)> { (FocusTest.Test, "Single focus test"),
            (BesselBeamTest.Test, "Bessel beam test"),
            (PlaneWaveTest.Test, "Plane wave test"),
            (WavTest.Test, "Wav modulation test"),
            (FocusSTMTest.Test, "FocusSTM test"),
            (GainSTMTest.Test, "GainSTM test"),
            (GainHoloTest.Test, "Multiple foci test"),
            (CustomTest.Test, "Custom Gain & Modulation test"),
            (FlagTest.Test, "Flag test"),
            (TransTest.Test, "TransducerTest test"),
            (GroupByTransducerTest.Test, "Group (by Transducer) test")
        };
        if (autd.Geometry.NumDevices >= 2) examples.Add((GroupByDeviceTest.Test, "Group (by Device) test"));

        Console.WriteLine("======== AUTD3 firmware information ========");
        Console.WriteLine(string.Join("\n", await autd.FirmwareInfoListAsync()));
        Console.WriteLine("============================================");

        while (true)
        {
            Console.WriteLine(string.Join("\n", examples.Select((example, i) => $"[{i}]: {example.Item2}")));
            Console.WriteLine("[Others]: finish");
            Console.Write("Choose number: ");

            if (!int.TryParse(Console.ReadLine(), out var idx) || idx >= examples.Count) break;

            var fn = examples[idx].Item1;
            await fn(autd);

            Console.WriteLine("press any key to finish...");
            Console.ReadKey(true);

            Console.WriteLine("finish.");
            await autd.SendAsync(new Stop());
        }

        await autd.CloseAsync();
    }
}
